collections/_article/progress-report-web-export-in-4-3.md

```diff
@@ -0,0 +1,298 @@
+
+However, it came with a downside too: it introduced a new, virtually unsolvable bug:
+
+The single-threaded build introduced garbles in the audio of games, making them unplayable on low-end machines such as
+laptops or phones.
+
+When I mean unplayable on low-end machines, I mean it.
+[Try it yourself.](https://adamscott.github.io/truck-town-demo-main-thread/){:target="_blank"}
+(Consider yourself lucky if it doesn't glitch out.)
+
+Yes. Web single-threaded audio stream playback with Godot 4.2 is that bad. It is marginally better on Godot 3.x, but the
+problem exist there as well.
+
+## Sample playback to the rescue!
+[Here comes my <abbr title="Pull Request">PR</abbr>](https://github.com/godotengine/godot/pull/91382) (not yet merged at
+the time of writing) for the rescue. [Here's the same demo without sound issues.](https://adamscott.github.io/truck-town-main-thread-samples/){:target="_blank"}
+
+### What are samples?
+Using samples for games is not a novel idea. In fact, for a long time, it was the _de facto_ way to play sound and music
+on game consoles. As they didn't have the processing power to do the mixing on the CPU, you could send audio clips
+(hence the term "sample") to a special chip and tell it how and when to play them.
+
+Since the audio chip was operating on it's own, the sound would never lag nor glitch out if the game is lagging.
+
+### (Re)introducing samples to Godot
+When open sourced, the Godot engine offered two ways to emit sound and music: streams and samples. Sample support was used for
+the Web exports and for platforms such as the <abbr title="PlayStation Portable">PSP</abbr>.
+
+![The long gone SampleLibrary resource and SamplePlayer2D node on Godot 2.1](/storage/blog/progress-report-web-export-in-4-3/sample-library.webp)
+
+But when Godot 3 released, it was decided that samples support was not needed anymore. Game consoles nowadays ask
+developers to mix themselves the audio using the CPU. And for the Web, we could mix the audio using a dedicated thread,
+like other platforms since `SharedArrayBuffer`s exist.
+
+This is why there's audio garbles in single-threaded builds (without my "fix"). Audio rendering is tied with the frame-
+rate. And if it drops, there's not enough audio frames to fill out the buffer, hence the glitches.
+
+What [my <abbr title="Pull Request">PR</abbr>](https://github.com/godotengine/godot/pull/85939) does is to reinstate
+sample playback. At least at the time of writing, for the Web platform only.
+
+### Near seamless integration
+Unfortunately, the way it was implemented in Godot 2.1, at least, was that you _had_ to choose between streams or samples.
+What I mean is that you had to choose between `AudioStreamNode`s or `AudioSampleNode`s. That meant custom code to handle
+which node to play, or even worse, separate projects for Web projects.
+
+Even worse, `AudioSampleNode`s had `SampleLibrary`s that you had to populate with samples separately. Not the easiest way
```

Comment:
It's `AudioSample`/`AudioStream` _resources_, and `SamplePlayer`/`StreamPlayer` _nodes_.